src/graph_ir/low_graph_souffle_emitter.cc

```diff
@@ -0,0 +1,48 @@
+//-----------------------------------------------------------------------------
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//-----------------------------------------------------------------------------
+
+#include <vector>
+#include <string>
+#include "low_graph_souffle_emitter.h"
+
+using namespace std;
+
+string LowGraphSouffleEmitter::EmitPathList(
+        vector<LowGraphAccessPath> pathList) {
+    string ret = "$Nil()";
+    for (auto path: pathList) {
+        ret = "$Cons(" + path.full_path_name + ", " + ret + ")";
```

Comment:
This is a place where you do not want to use string concatenation. While the `+` operator on strings is fine in a leaf method when the number of concatenations is statically known, doing it `N` times for some input parameter `N` results in `N^2` performance. Use `ostringstream` instead. Similar throught this file.

Reply:
This code needs to put text at the beginning of the previous version of the string, which I think requires a copy per iteration even if a stringstream is used. Am I mistaken? 

For the rest, I'm now using stringstreams.

Reply:
You raise a good point, and that makes it so that `stringstream` is not as easy to drop in. Rather than repeatedly nesting, I'd say just construct the prefix and the suffix separately as follows:

std::stringstream result;
for (auto path: pathList) {
  result << "$Cons(" << path.full_path_name << ", ";
}
result << "$Nil()";
for (int i = 0; i < pathList.size(); ++i) {
  result << ")";
}

This will run in `O(N)` time, where `N` is the number of path elements.